### PR TITLE
Linux texture support

### PR DESCRIPTION
**Pull request by anirudhb**
## Description

Add support for external textures to the Linux platform shell.

## Related Issues

Fixes flutter/flutter#64188.

## Tests

I added the following tests:
- [x] Tests for `FlTextureRegistrar`
- [x] Tests for `FlExternalTextureGl`

## Checklist

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by stuartmorgan** [comment]
Filed https://github.com/flutter/flutter/issues/64576 for the lint issue; we'll figure out the plan to resolve the unrelated issues there. I'll review the rest for functionality now.

**Comment by anirudhb** [comment]
@stuartmorgan Resolved nits and added partial tests. Re-review?

**Comment by stuartmorgan** [comment]
> Filed [flutter/flutter#64576](https://github.com/flutter/flutter/issues/64576) for the lint issue; we'll figure out the plan to resolve the unrelated issues there.

Based on https://github.com/flutter/flutter/issues/64576, the solution for now should be to add `// FLUTTER_NOLINT` just after the license header. It's unclear how the file ended up not having it already, but you can just add it to this PR.

**Comment by stuartmorgan** [comment]
In the future, please don't force-push to in-review PRs. It removes the ability to use very helpful review features like seeing everything that changed since the last review.

File: shell/platform/linux/BUILD.gn [hunk 1]
```diff
@@ -80,6 +81,7 @@ source_set("flutter_linux_sources") {
     "fl_binary_messenger.cc",
     "fl_dart_project.cc",
     "fl_engine.cc",
+    "fl_external_texture_gl.cc",
```
Comment (stuartmorgan) [inline shell/platform/linux/BUILD.gn:84]: I don't see a test file for this new class.

File: shell/platform/linux/testing/fl_test.cc [hunk 2]
```diff
@@ -39,3 +47,15 @@ gchar* bytes_to_hex_string(GBytes* bytes) {
     g_string_append_printf(hex_string, "%02x", data[i]);
   return g_string_free(hex_string, FALSE);
 }
+
+// Creates a mock engine that responds to platform messages.
```
Comment (stuartmorgan) [inline shell/platform/linux/testing/fl_test.cc:51]: Declaration comments should not be repeated at the definition.

File: shell/platform/linux/testing/fl_test.h [hunk 3]
```diff
@@ -18,6 +20,9 @@ GBytes* hex_string_to_bytes(const gchar* hex_string);
 // Helper function to convert GBytes into a hexadecimal string (e.g. "01feab")
 gchar* bytes_to_hex_string(GBytes* bytes);
 
+// Creates a mock engine that responds to platform messages.
```
Comment (stuartmorgan) [inline shell/platform/linux/testing/fl_test.h:23]: This description doesn't seem adequate outside its old much more limited context. Please elaborate on what is and is not mocked in this engine. (It looks like it may just be the renderer?)

File: shell/platform/linux/fl_external_texture_gl.cc [hunk 4]
```diff
@@ -0,0 +1,133 @@
+// Copyright 2020 The Flutter Authors. All rights reserved.
```
Comment (stuartmorgan) [inline shell/platform/linux/fl_external_texture_gl.cc:1]: All engine code uses 2013.

File: shell/platform/linux/fl_external_texture_gl.cc [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+                     GLint level,
+                     GLint internalformat,
+                     GLsizei width,
+                     GLsizei height,
+                     GLint border,
+                     GLenum format,
+                     GLenum type,
+                     const void* data);
+  void (*deleteTextures)(GLsizei n, const GLuint* textures);
+} gl;
```
Comment (stuartmorgan) [inline shell/platform/linux/fl_external_texture_gl.cc:26]: Please use the construction I described in the Windows discussion I linked to before; it makes it much clearer where initialization happens.

File: shell/platform/linux/fl_texture_registrar_test.cc [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+#include "flutter/shell/platform/embedder/embedder.h"
+#include "flutter/shell/platform/linux/fl_external_texture_gl.h"
+#include "flutter/shell/platform/linux/fl_texture_registrar_private.h"
+#include "flutter/shell/platform/linux/testing/fl_test.h"
+#include "gtest/gtest.h"
+
+#include <gmodule.h>
+
+// Implements eglGetProcAddress but returns dummy functions.
+// static void* eglGetProcAddress(const char* name) {
```
Comment (stuartmorgan) [inline shell/platform/linux/fl_texture_registrar_test.cc:15]: There's a bunch of commented out code in this PR, which isn't something we would land. I'm not clear what the status here is; is finishing the tests blocked on something? Your request for re-review didn't mention anything.

File: shell/platform/linux/fl_texture_registrar_test.cc [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+TEST(FlTextureRegistrarTest, MarkTextureFrameAvailable) {
+  g_autoptr(FlEngine) engine = make_mock_engine();
+  FlTextureRegistrar* registrar = fl_texture_registrar_new(engine);
+  int64_t id =
+      fl_texture_registrar_register_texture(registrar, nullptr, nullptr);
+  fl_texture_registrar_mark_texture_frame_available(registrar, id);
+}
+
+// TODO(anirudhb): Re-enable after figuring out how to stub out
+// eglGetProcAddress. Also write tests for #FlExternalTextureGl as well.
```
Comment (stuartmorgan) [inline shell/platform/linux/fl_texture_registrar_test.cc:45]: If it's this, you can make the resolver something that is provided in the constructor for the object, and pass a different resolver in unit tests.

File: shell/platform/linux/fl_engine_private.h [hunk 8]
```diff
@@ -175,6 +175,41 @@ GBytes* fl_engine_send_platform_message_finish(FlEngine* engine,
                                                GAsyncResult* result,
                                                GError** error);
 
+/**
+ * fl_engine_mark_texture_frame_available:
+ * @engine: an #FlEngine.
+ * @texture_id: an int64_t.
```
Comment (robert-ancell) [inline shell/platform/linux/fl_engine_private.h:181]: This should document what `texture_id` represents, e.g. 'the identifier of the texture whose frame has been updated'. (Same for cases below).

File: shell/platform/linux/fl_engine_private.h [hunk 9]
```diff
@@ -175,6 +175,41 @@ GBytes* fl_engine_send_platform_message_finish(FlEngine* engine,
...
+ *
+ * Tells the Flutter engine that a new texture frame is available for the given
+ * texture.
+ *
+ * Returns: true on success.
+ */
+bool fl_engine_mark_texture_frame_available(FlEngine* engine,
```
Comment (robert-ancell) [inline shell/platform/linux/fl_engine_private.h:188]: For consistency with the existing code this should probably return gboolean, and have the documentation "Returns: %TRUE on success".

File: shell/platform/linux/fl_external_texture_gl.cc [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+struct _FlExternalTextureGl {
+  GLuint gl_texture_id;
```
Comment (robert-ancell) [inline shell/platform/linux/fl_external_texture_gl.cc:29]: This is missing `GObject parent_instance` as the first element in this struct - the current code will not create a valid object.

File: shell/platform/linux/fl_external_texture_gl.cc [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+static void fl_external_texture_gl_class_init(FlExternalTextureGlClass* klass) {
+  G_OBJECT_CLASS(klass)->dispose = fl_external_texture_gl_dispose;
+}
+
+static void fl_external_texture_gl_init(FlExternalTextureGl* self) {}
+
+int64_t fl_external_texture_gl_texture_id(FlExternalTextureGl* self) {
+  return reinterpret_cast<int64_t>(self);
```
Comment (robert-ancell) [inline shell/platform/linux/fl_external_texture_gl.cc:52]: Should be more like:

```
g_return_val_if_fail(GL_IS_EXTERNAL_TEXTURE_GL(self), -1);
return self->gl_texture_id;
```

File: shell/platform/linux/fl_external_texture_gl.cc [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+bool fl_external_texture_gl_populate_texture(
+    FlExternalTextureGl* self,
+    size_t width,
+    size_t height,
+    FlutterOpenGLTexture* opengl_texture) {
```
Comment (robert-ancell) [inline shell/platform/linux/fl_external_texture_gl.cc:59]: Add guards: `g_return_val_if_fail(GL_IS_EXTERNAL_TEXTURE_GL(self), FALSE)`

(and other cases below).

File: shell/platform/linux/public/flutter_linux/fl_texture_registrar.h [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+ */
+typedef struct {
+  const uint8_t* buffer;
+  size_t width;
+  size_t height;
+} FlPixelBuffer;
+
+/**
+ * The pixel buffer copy callback provided to the Flutter engine to copy the
+ * texture.
```
Comment (robert-ancell) [inline shell/platform/linux/public/flutter_linux/fl_texture_registrar.h:28]: Please document the parameters in the callback.

File: shell/platform/linux/public/flutter_linux/fl_texture_registrar.h [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+#include <glib-object.h>
+#include <stdint.h>
+
+G_BEGIN_DECLS
+
+/**
+ * An image buffer object.
+ */
+typedef struct {
+  const uint8_t* buffer;
```
Comment (robert-ancell) [inline shell/platform/linux/public/flutter_linux/fl_texture_registrar.h:21]: Please document these fields.

File: shell/platform/linux/public/flutter_linux/fl_texture_registrar.h [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+  size_t width;
+  size_t height;
+} FlPixelBuffer;
+
+/**
+ * The pixel buffer copy callback provided to the Flutter engine to copy the
+ * texture.
+ */
+typedef const FlPixelBuffer* (*FlTextureCallback)(size_t width,
```
Comment (robert-ancell) [inline shell/platform/linux/public/flutter_linux/fl_texture_registrar.h:30]: You could consider removing `FlPixelBuffer` and do something like:

```
typedef gboolean (*FlTextureCallback)(size_t* width, size_t* height, uint8_t* buffer)
```

Where `width` and `height` are inout parameters, and `buffer` is an out parameter. I say this because its a little unusual to use structs like this is GObject code. If the callback can fail it would also be worth having a `GError**` parameter to allow the callback implementation to report an error that occurred.

File: shell/platform/linux/public/flutter_linux/fl_texture_registrar.h [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+/**
+ * FlTextureRegistrar:
+ *
+ * #FlTextureRegistrar is used when registering textures.
+ */
+
+/**
+ * fl_texture_registrar_register_texture:
+ * @registrar: an #FlTextureRegistrar.
+ * @texture: an #FlTexture.
```
Comment (robert-ancell) [inline shell/platform/linux/public/flutter_linux/fl_texture_registrar.h:49]: `texture_callback` / `user_data` not documented. `texture` not a parameter.

**Review by robert-ancell — CHANGES_REQUESTED** [review]
Thanks for working on this!

**Comment by anirudhb** [comment]
I've addressed nits. Finished tests as well by changing the mock implementation in `mock_egl.cc`.

File: shell/platform/linux/fl_external_texture_gl.cc (hunk 11 above)
Comment (anirudhb) [reply shell/platform/linux/fl_external_texture_gl.cc:73]: I did not end up returning the OpenGL texture ID since its value is always zero before `fl_external_texture_gl_copy_pixel_buffer` is called, so we end up always passing zero to the embedder.

**Comment by stuartmorgan** [comment]
The conversation starting [here](https://github.com/flutter/engine/pull/19405#issuecomment-684129332) is relevant to this PR as well; we should address them the same way, ideally.

**Comment by anirudhb** [comment]
I've read that conversation, and my personal thinking is that this initial functionality is worth merging, then we can handle more efficient texture handoff later. Like Windows, Linux doesn't have any graphics API-independent way to hand off textures so I think this is the best way for now, since it abstracts that functionality from the user.

**Comment by stuartmorgan** [comment]
> this initial functionality is worth merging, then we can handle more efficient texture handoff later

As I explained there, it's important that enough thought is given to things like API names in this version that the more efficient version(s) can be added later without it being a breaking change (or being stuck with confusing APIs where there are 2+ variants, but one of them has a confusingly generic name).

File: shell/platform/linux/fl_external_texture_gl.h [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+ * @texture_callback: An #FlTextureCallback.
+ * @user_data: A void*.
+ *
+ * Creates a new #FlExternalTextureGl.
+ *
+ * Returns: a new #FlExternalTextureGl.
+ */
+FlExternalTextureGl* fl_external_texture_gl_new(
+    FlTextureCallback texture_callback,
+    void* user_data);
```
Comment (anirudhb) [inline shell/platform/linux/fl_external_texture_gl.h:38]: This should definitely be part of the interface but not sure how to make it generic based on `FlTextureCallback` returning different values. One way would be to change the output buffer of `FlTextureCallback` to a `void*`, then the subclass defines what is supposed to be output.

File: shell/platform/linux/fl_texture_registrar_private.h [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+ * texture identified by |texture_id|.
+ *
+ * Returns true on success.
+ */
+bool fl_texture_registrar_populate_texture(
+    FlTextureRegistrar* registrar,
+    int64_t texture_id,
+    size_t width,
+    size_t height,
+    FlutterOpenGLTexture* opengl_texture);
```
Comment (anirudhb) [inline shell/platform/linux/fl_texture_registrar_private.h:42]: Again, the `FlutterOpenGLTexture` should be changed to a `void*`.

File: shell/platform/linux/public/flutter_linux/fl_texture_registrar.h [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+ * texture.
+ * @width: inout parameter representing the given width and the actual width.
+ * @height: inout parameter representing the given height and the actual height.
+ * @buffer: out parameter representing the pixel buffer.
+ *
+ * Returns: %TRUE on success.
+ */
+typedef gboolean (*FlTextureCallback)(size_t* width,
+                                      size_t* height,
+                                      const uint8_t** buffer,
```
Comment (anirudhb) [inline shell/platform/linux/public/flutter_linux/fl_texture_registrar.h:28]: This might be better redefined as `void* out`, so it is agnostic of the underlying output and can therefore be safely passed to all `FlExternalTexture` subclasses.

File: shell/platform/linux/public/flutter_linux/fl_texture_registrar.h [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+                     fl_texture_registrar,
+                     FL,
+                     TEXTURE_REGISTRAR,
+                     GObject)
+
+/**
+ * FlTextureRegistrar:
+ *
+ * #FlTextureRegistrar is used when registering textures.
+ */
```
Comment (anirudhb) [inline shell/platform/linux/public/flutter_linux/fl_texture_registrar.h:41]: Internally, the `FlTextureRegistrar` hashmap should be managing `FlExternalTexture`s not their subclasses.

File: shell/platform/linux/public/flutter_linux/fl_texture_registrar.h [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+ * @user_data: user data to pass to the callback.
+ *
+ * Registers a texture callback and returns the ID for that texture.
+ *
+ * Returns: an int64_t.
+ */
+int64_t fl_texture_registrar_register_texture(
+    FlTextureRegistrar* registrar,
+    FlTextureCallback texture_callback,
+    void* user_data);
```
Comment (anirudhb) [inline shell/platform/linux/public/flutter_linux/fl_texture_registrar.h:56]: On making this agnostic-- it might make sense to make an enum here to differentiate which type of texture is being registered (i.e. `OpenGL`, `Vulkan` or `Pixel Buffer` which should be agnostic). Then the texture registrar would instantiate the respective underlying `FlExternalTexture` subclass. Managing `Pixel Buffer` callbacks could be done by the texture registrar itself, maybe by simply wrapping it with another callback.

**Review by anirudhb — COMMENTED** [review]
@stuartmorgan I've left some notes on what I think could be a way to make this API more agnostic of the underlying texture backend. Please comment any thoughts or nits you may have on this design.

File: shell/platform/linux/testing/fl_test.cc [hunk 22]
```diff
@@ -39,3 +48,14 @@ gchar* bytes_to_hex_string(GBytes* bytes) {
     g_string_append_printf(hex_string, "%02x", data[i]);
   return g_string_free(hex_string, FALSE);
 }
+
+FlEngine* make_mock_engine() {
```
Comment (robert-ancell) [inline shell/platform/linux/testing/fl_test.cc:52]: Can you split this change out into a separate PR?

Comment (robert-ancell) [reply shell/platform/linux/testing/fl_test.cc:52]: There's a number of cases where this could be common.

File: shell/platform/linux/fl_engine.cc [hunk 23]
```diff
@@ -210,6 +213,20 @@ static bool fl_engine_gl_make_resource_current(void* user_data) {
   return result;
 }
 
+static bool fl_engine_gl_external_texture_frame_callback(
```
Comment (robert-ancell) [inline shell/platform/linux/fl_engine.cc:216]: Needs a comment above this function.

File: shell/platform/linux/fl_engine.cc [hunk 24]
```diff
@@ -571,8 +593,37 @@ void fl_engine_send_mouse_pointer_event(FlEngine* self,
   FlutterEngineSendPointerEvent(self->engine, &fl_event, 1);
 }
 
+gboolean fl_engine_mark_texture_frame_available(FlEngine* self,
+                                                int64_t texture_id) {
+  g_return_val_if_fail(FL_IS_ENGINE(self), FALSE);
+  return FlutterEngineMarkExternalTextureFrameAvailable(self->engine,
+                                                        texture_id) == kSuccess
+             ? TRUE
```
Comment (robert-ancell) [inline shell/platform/linux/fl_engine.cc:601]: `== kSuccess` is sufficient to return a boolean, i.e. don't need the `? TRUE : FALSE`. (Same below).

File: shell/platform/linux/fl_external_texture_gl.cc (hunk 11 above)
Comment (robert-ancell) [reply shell/platform/linux/fl_external_texture_gl.cc:73]: Ah, I understand now.

**Comment by stuartmorgan** [comment]
> @stuartmorgan I've left some notes on what I think could be a way to make this API more agnostic of the underlying texture backend. Please comment any thoughts or nits you may have on this design.

Some of https://github.com/flutter/engine/pull/19405#pullrequestreview-497943610 (the C API layer part) is relevant here.

You've suggested `void*` in several places; that's a solution of absolute last resort because it means we're no longer providing any type safety for that portion of the API. I don't see any fundamental reason that a discriminated-union-style approach, just as is used for the embedder.h texture API, couldn't be used here to allow for a set of possible textures (this would stand in for the `std::variant` at the C++ wrapper layer in my comments on the Windows version).

Or maybe there's something that's a higher-level `std::variant`-like tool in GObject that would be useful instead of a discriminated union? @robert-ancell ?

File: shell/platform/linux/testing/fl_test.cc (hunk 22 above)
Comment (anirudhb) [reply shell/platform/linux/testing/fl_test.cc:52]: Done, #21585.

**Comment by chinmaygarde** [comment]
From PR triage, is there any progress being made on this PR? If not, can we close this as it looks stale (last update 21 days ago).

**Comment by anirudhb** [comment]
I do plan to make progress on this PR, but I'm waiting on #21585 being merged first to avoid conflicts. Additionally, this and #19405 both seem to be stalled on working out a proper API design, so it would be nice to keep it consistent across both platforms.

**Comment by stuartmorgan** [comment]
>  Additionally, this and #19405 both seem to be stalled on working out a proper API design

I outlined a path forward in 19405 for the API.

**Comment by anirudhb** [comment]
Looks like this PR will need quite a bit of refactoring to pass review again. I don't think I'll have time for the next month or so, so I've left this as a draft for now. I am definitely still interested in adding support, and will undraft the PR when I feel it's ready :)

**Comment by chinmaygarde** [comment]
Looks like progress on this has stalled. Closing it for now. Please re-open when needed.

**Comment by stuartmorgan** [comment]
Please read the recent comments; it's not ready for final review yet, so can't be merged as-is.

**Comment by huanghongxun** [comment]
I have redesigned API and resubmit this PR https://github.com/flutter/engine/pull/24916.
Please have a look.
